### PR TITLE
Updated formatting for ml-frameworks

### PR DESCRIPTION
**Pull request by DivyaAmirtharaj**

### CODE REVIEW

File: frameworks.qmd [hunk 1]
```diff
@@ -1,119 +1,1785 @@
 # AI Frameworks
```
Comment (BaeHenryS) [inline frameworks.qmd:1]: Leaving Comment to Test Co-Authorship

**Review by BaeHenryS — CHANGES_REQUESTED** [review]

**Comment by sophiacho1** [comment]
Testing

File: frameworks.qmd [hunk 2]
```diff
@@ -1,119 +1,1785 @@
...
 
-::: {.callout-tip collapse="true"}
-## Learning Objectives
```
Comment (sophiacho1) [inline frameworks.qmd:4]: Testing

**Comment by profvjreddi** [comment]
Great job on pulling this draft together @DivyaAmirtharaj @sophiacho1 @BaeHenryS and Emeka. 

Let the feedback begin! 😄

File: frameworks.qmd [hunk 3]
```diff
@@ -1,119 +1,1785 @@
...
+enable scalable machine learning from the cloud to the edge.
+
+## Framework Evolution
+
+Machine learning frameworks have evolved significantly over time to meet
+the diverse needs of machine learning practitioners and advancements in
+AI techniques. A few decades ago, building and training machine learning
+models required extensive low-level coding and infrastructure. Machine
+learning frameworks have evolved considerably over the past decade to
+meet the expanding needs of practitioners and rapid advances in deep
```
Comment (oishib) [inline frameworks.qmd:80]: This paragraph could be reorganized to be clearer and shorter. For example, these two sentences are very similar and could be condensed: "Machine learning frameworks have evolved significantly over time to meet the diverse needs of machine learning practitioners and advancements in AI techniques." and "Machine learning frameworks have evolved considerably over the past decade to meet the expanding needs of practitioners and rapid advances in deep learning techniques"

File: frameworks.qmd [hunk 4]
```diff
@@ -1,119 +1,1785 @@
...
+learning techniques. Early neural network research was constrained by
+insufficient data and compute power. Building and training machine
+learning models required extensive low-level coding and infrastructure.
+But the release of large datasets like ImageNet in 2009 and advancements
+in parallel GPU computing unlocked the potential for far deeper neural
+networks.
+
+The first ML frameworks, Theano (2007) and Caffe (2014), were developed
```
Comment (oishib) [inline frameworks.qmd:88]: Are these citations properly linked to corresponding BibTex entries?

File: frameworks.qmd [hunk 5]
```diff
@@ -1,119 +1,1785 @@
...
+low-level frameworks do not have. Thus, high-level frameworks may be
+easier to use, but are not as customizable as low-level frameworks (i.e.
+users of low-level frameworks can define custom layers, loss functions,
+optimization algorithms, etc.). Examples of high-level frameworks
+include TensorFlow/Keras and PyTorch. Examples of low-level ML
+frameworks include TensorFlow with low-level APIs, Theano, Caffe,
+Chainer, and CNTK.
+
+Frameworks like Theano and Caffe used static computational graphs which
+required rigidly defining the full model architecture upfront. Static
```
Comment (oishib) [inline frameworks.qmd:114]: It might help to add brief explanations of what a computational graph is, what you would be declaring upfront, whether "on-the-fly" means "during training" or "during inference" or just "during runtime," etc.

File: frameworks.qmd [hunk 6]
```diff
@@ -1,119 +1,1785 @@
...
+graphs require upfront declaration and limit flexibility. Dynamic graphs
+construct on-the-fly for more iterative development. But around 2016,
+frameworks began adopting dynamic graphs like PyTorch and TensorFlow 2.0
+which can construct graphs on-the-fly. This provides greater flexibility
```
Comment (alex-oesterling) [inline frameworks.qmd:118]: "Dynamic graphs
construct on-the-fly for more iterative development. But around 2016,
frameworks began adopting dynamic graphs like PyTorch and TensorFlow 2.0
which can construct graphs on-the-fly." 

These sentences are difficult to read, I would suggest reversing them + some tweaks: "But around 2016, frameworks such as Pytorch and TF 2.0 began adopting dynamic graphs which are constructed on-the-fly for more iterative development."

File: frameworks.qmd [hunk 7]
```diff
@@ -1,119 +1,1785 @@
...
+training epochs reduces bias and improves generalization.
+
+Data loaders also support caching and prefetching strategies to optimize
+data delivery for fast, smooth model training. Caching preprocessed
+batches in memory so they can be reused efficiently during multiple
+training steps. Caching these batches in memory eliminates redundant
+processing. Prefetching, on the other hand, involves preloading
+subsequent batches, ensuring that the model never idles waiting for
+data.
+
```
Comment (ushabhalla) [inline frameworks.qmd:723]: This data loaders section is great, but I think it can be consolidated a bit as some parts are a bit repetitive and it was split up slightly un-intuitively. From what I can tell the main points are that dataloaders (1) read the data from disk or memory, (2) allow for parallel computation on GPU via batching, (3) allow for shuffling (and I think augmentation should be mentioned in this part even though it is the next section), and (4) cacheing for efficiency. I think that points 1 and 4 can be consolidated, as well as 2 and 3 to simply say that dataloaders handle the fetching and cacheing of the inputs to the model, and can be used to batch, shuffle, and augment data before sending it through the model.

File: frameworks.qmd [hunk 8]
```diff
@@ -1,119 +1,1785 @@
...
+-   Activation grids - Illustrate features learned by convolutional
+    > filters.
+
+-   Projection - Reduce dimensionality for intuitive visualization.
+
+-   Precision-recall curves - Assess classification tradeoffs.
+
+Tools like TensorBoard (TensorFlow) and TensorWatch (PyTorch) enable
+real-time metrics and visualization during training.
```
Comment (ushabhalla) [inline frameworks.qmd:894]: Weights and Biases may also be a good reference for this! [W&B link](https://wandb.ai/site)

File: frameworks.qmd [hunk 9]
```diff
@@ -1,119 +1,1785 @@
...
+
+Training a neural network is fundamentally an iterative process that
+seeks to minimize a loss function. At its core, the goal is to fine-tune
+the model weights and parameters to produce predictions as close as
+possible to the true target labels. Machine learning frameworks have
+greatly streamlined this process by offering extensive support in three
+critical areas: loss functions, optimization algorithms, and
+regularization techniques.
+
+Loss Functions are useful to quantify the difference between the
```
Comment (jared-ni) [inline frameworks.qmd:756]: It could be great to illustrate this with the image of a common mathematical loss function and explain the different components. It would strengthen this paragraph and make it more informative (since we probably shouldn't assume too much prior knowledge in a potential all-in-one textbook, though the topic may seem simple to people in the field).

File: frameworks.qmd [hunk 10]
```diff
@@ -1,119 +1,1785 @@
...
+computes how much changes tuning the weights in a certain way changes
+the value of the loss function. Knowing this gradient, the model moves
+in the direction that reduces the gradient. There are many challenges
+associated with this, however, primarily stemming from the fact that the
+optimization problem is not convex, making it very easy to solve, and
+more details about this will come in the AI training sections. Modern
+frameworks come equipped with efficient implementations of several
+optimization algorithms, many of which are variants of gradient descent
+algorithms with stochastic methods and adaptive learning rates. More
+information can be found in the AI training section.
```
Comment (jared-ni) [inline frameworks.qmd:775]: It helps to illustrate the loss function and algorithms like gradient descent stochastic methods or learning rates with some equations. Again, I think this part is assuming too much prior knowledge, and those who are not familiar with the buzz words might not understand this paragraph.

File: frameworks.qmd [hunk 11]
```diff
@@ -1,119 +1,1785 @@
...
+
+Last but not least, overly complex models tend to overfit, meaning they
+perform well on the training data but fail to generalize to new, unseen
+data (see Overfitting). To counteract this, regularization methods are
+employed to penalize model complexity and encourage it to learn simpler
+patterns. Dropout for instance randomly sets a fraction of input units
+to 0 at each update during training, which helps prevent overfitting.
```
Comment (jared-ni) [inline frameworks.qmd:782]: Since the paragraph explains overfitting, wouldn't it also be helpful to touch on underfitting, and in what situations either case may be encountered? I think it would be informative to add examples with potential graphs to illustrate this point.

File: frameworks.qmd [hunk 12]
```diff
@@ -1,119 +1,1785 @@
...
+### Model Training Support
+
+Before training a defined neural network model, a compilation step is
+required. During this step, the high-level architecture of the neural
+network is transformed into an optimized, executable format. This
+process comprises several steps. The construction of the computational
+graph is the first step. It represents all the mathematical operations
+and data flow within the model. We discussed this earlier.
+
+During training, the focus is on executing the computational graph.
```
Comment (jared-ni) [inline frameworks.qmd:798]: Computational graph might be familiar to people who are exposed to AI/ML, but it doesn't mean much to people who aren't experienced in the field. Since this textbook covers this topic, I think a graphical representation of a computational graph would be good to include here!

File: frameworks.qmd [hunk 13]
```diff
@@ -1,119 +1,1785 @@
...
+distributing the models into personal devices for them to be trained on
+device. At the beginning, a base global model is trained on a central
+server to be distributed to all devices. Using this base model, the
+devices individually compute the gradients and send them back to the
+central hub. Intuitively this is the transfer of model parameters
+instead of the data itself. This innovative approach allows the model to
+be trained with many different datasets (which, in our example, would be
+the set of images that are on personal devices), without the need to
+transfer a large amount of potentially sensitive data. However,
+federated learning also comes with a series of challenges.
```
Comment (jared-ni) [inline frameworks.qmd:1126]: Great paragraph on Federated Learning! I think a graph to illustrate how federated learning works might also be cool.

File: frameworks.qmd [hunk 14]
```diff
@@ -1,119 +1,1785 @@
...
+Microcontrollers expand possibilities. They provide greater flexibility
+for custom and advanced models beyond the core operators in TFLite
+Micro. However, this comes at the cost of a larger memory footprint.
+These frameworks are ideal for automotive systems, drones, and more
+powerful edge devices that can benefit from greater model
+sophistication.
+
+Frameworks specifically built for specialized hardware like CMSIS-NN on
+Cortex-M processors can further maximize performance, but sacrifice
+portability. Integrated frameworks from processor vendors tailor the
```
Comment (jared-ni) [inline frameworks.qmd:1773]: Great summary of the tradeoffs when using these frameworks!

**Review by jared-ni — COMMENTED** [review]
Awesome text!

File: frameworks.qmd [hunk 15]
```diff
@@ -1,119 +1,1785 @@
...
+frameworks to focus on high-value tasks like model architecture, feature
+engineering, and hyperparameter tuning instead of infrastructure. The
+goal is to efficiently build and deploy performant models that solve
+real-world problems.
+
+In this chapter, we will explore today\'s leading cloud frameworks and
+how they have adapted models and tools specifically for embedded and
+edge deployment. We will compare programming models, supported hardware,
+optimization capabilities, and more to fully understand how frameworks
+enable scalable machine learning from the cloud to the edge.
```
Comment (arnaumarin) [inline frameworks.qmd:71]: Overall the introduction is very clear providing a comprehensive overview of the ML frameworks, utilities and key features + it sets a clear scope for the chapter, which i find very useful in general.

The only thing I feel (perhaps is too personal) it lacks is a specific mention of the challenges of running ML models on resource-constrained devices, which would be crucial for a TinyML-focused book.  For example, in the first paragraph you introduce the TF, torch, and "specialized frameworks for
embedded devices". Why don't you just mention Lite/Lite Micro and more here? 
You guys might also want to add a sentence about how the constraints of TinyML (power, memory, compute) necessitate different framework capabilities or optimizations. (After this, I know you mention it with TF Lite, Lite Micro, etc.)

File: frameworks.qmd [hunk 16]
```diff
@@ -1,119 +1,1793 @@
...
+
+Popularity of ML frameworks in the United States as measured by Google
+web searches
+
+However, a one-size-fits-all approach does not work well across the
+spectrum from cloud to tiny edge devices. Different frameworks represent
+various philosophies around graph execution, declarative versus
+imperative APIs, and more. Declarative defines what the program should
+do while imperative focuses on how it should do it step-by-step. For
+instance, TensorFlow uses graph execution and declarative-style modeling
```
Comment (AditiR-42) [inline frameworks.qmd:168]: It would be helpful to provide more explanation on the difference between declarative and imperative APIs. Consider giving an example of the actual function of a declarative API vs an imperative API instead of just citing TensorFlow and Pytorch.

File: frameworks.qmd [hunk 17]
```diff
@@ -1,119 +1,1793 @@
...
+require the initialization of new parameters or the removal of existing
+ones, complicating the management and synchronization tasks of the
+parameter servers. This made it harder to implement models outside the
+neural framework or models that required dynamic computation graphs.
+
+TensorFlow was designed to be a more general computation framework[^2] where
+the computation is expressed as a data flow graph. This allows for a
+wider variety of machine learning models and algorithms outside of just
+neural networks, and provides flexibility in refining models.
+
```
Comment (AditiR-42) [inline frameworks.qmd:303]: I think it could be helpful to give more context on how exactly a dynamic computation graph works! This reference might be useful: https://www.geeksforgeeks.org/dynamic-vs-static-computational-graphs-pytorch-and-tensorflow/#

File: frameworks.qmd [hunk 18]
```diff
@@ -1,119 +1,1793 @@
...
+begin to overfit (that is, if its performance on the validation set
+declines), training is automatically halted, conserving computational
+resources and time.
+
+ML frameworks incorporate a blend of model compilation, enhanced batch
+processing methods, and utilities such as checkpointing and early
+stopping. These resources manage the complex aspects of performance,
+enabling practitioners to zero in on model development and training. As
+a result, developers experience both speed and ease when utilizing the
+capabilities of neural networks.
```
Comment (AditiR-42) [inline frameworks.qmd:836]: It would be helpful to have a diagram of batch processing here so the reader can visualize how batch processing works. For example: https://media.geeksforgeeks.org/wp-content/uploads/20221020142629/Batchoperatingsystem1-660x330.png

File: frameworks.qmd [hunk 19]
```diff
@@ -1,119 +1,1793 @@
...
+proven to be a useful structure to efficiently compute matrix products
+and other operations within neural networks (such as convolutions).
+
+While TPU's can drastically reduce training times, it also has
+disadvantages. For example, many operations within the machine learning
+frameworks (primarily TensorFlow here since the TPU directly integrates
+with it) are not supported with the TPU's. It also cannot support custom
+custom operations from the machine learning frameworks, and the network
+design must closely align to the hardware capabilities.
+
```
Comment (AditiR-42) [inline frameworks.qmd:975]: It would be nice to add a more clear comparison between GPUs and TPUs here. What are the tradeoffs of using GPUs vs. TPUs and how does this change depending on use case?

File: frameworks.qmd [hunk 20]
```diff
@@ -1,119 +1,1785 @@
...
+graph frameworks. High-level frameworks provide a higher level of
+abstraction than low-level frameworks. That is, high-level frameworks
+have pre-built functions and modules for common ML tasks, such as
+creating, training, and evaluating common ML models as well as
+preprocessing data, engineering features, and visualizing data, which
+low-level frameworks do not have. Thus, high-level frameworks may be
+easier to use, but are not as customizable as low-level frameworks (i.e.
+users of low-level frameworks can define custom layers, loss functions,
+optimization algorithms, etc.). Examples of high-level frameworks
+include TensorFlow/Keras and PyTorch. Examples of low-level ML
```
Comment (sjohri20) [inline frameworks.qmd:109]: Does Pytorch not fall under "low-level frameworks" according to the above definition? It is possible to define custom layers, loss functions, optimizing algorithms in pytorch

File: frameworks.qmd [hunk 21]
```diff
@@ -1,119 +1,1785 @@
...
-### Hardware
+Training in machine learning usually assumes the data to be identically
+and independently distributed. However, the nature of federated learning
+makes it so that this is not necessarily the case. In our example,
+devices may have similar photos from being at the same location, and the
+types of the cameras that each of the devices have may be different.
+There are many different techniques to compensate for this, such as
+adding a proximal term to achieve a balance between the local and global
+model, and adding a frozen [[global hypersphere
+classifier]{.underline}](https://arxiv.org/abs/2207.09413).
```
Comment (sjohri20) [inline frameworks.qmd:1139]: Labeled Data is Not Always Naturally Available:
- In many real-world scenarios, data collected on devices might not come with appropriate labels.
- Additionally, users, being a source of data, can be unreliable, meaning that even if data is labeled, there's no guarantee of its accuracy or relevance.

Non-IID Data (covered already, but can be made clearer):
- Each user's data is unique, leading to significant variance in the kind of data generated by different users.
- The unbalanced nature of data, where some users produce more data than others can affect the global model's performance.

Massively Distributed:
- The number of mobile device owners can be significantly larger than the average number of training samples on each device. This leads to a significant communication overhead.

Limited Communication:
- Mobile networks, which are typically used for such communication, can be unstable. This instability can lead to delayed or failed transmission of model updates, affecting the overall training process.

Heterogeneous Device/Silo Resources:
- Devices participating in FL can have varying computational powers and memory capacities, making it challenging to design algorithms that are efficient across all devices.

Privacy, Security, and Trust:
- Malicious clients or servers can attempt to reverse-engineer model updates to infer information about local data.
- With malicious clients or servers in the system, there's a risk of model poisoning or other types of adversarial attacks.

Comment (profvjreddi) [reply frameworks.qmd:1139]: That's excellent feedback. Thanks for providing that.

File: frameworks.qmd (hunk 4 above)
Comment (profvjreddi) [reply frameworks.qmd:88]: I believe @DivyaAmirtharaj @sophiacho1 @BaeHenryS Emeka are still working on those.

File: frameworks.qmd (hunk 6 above)
Comment (profvjreddi) [reply frameworks.qmd:118]: Thanks for providing this suggestion on how to rework things. That makes things so much easier to integrate.

File: frameworks.qmd (hunk 8 above)
Comment (profvjreddi) [reply frameworks.qmd:897]: That's a great suggestion, we should definitely add that.

File: frameworks.qmd (hunk 10 above)
Comment (profvjreddi) [reply frameworks.qmd:778]: Good point. 

There will be an AI training chapter that will be coming in before this chapter; we are working on it (just not within the class context), and our hope is that that chapter would have set up the necessary background that would then flow into this chapter. but good feedback though. Thank you.

File: frameworks.qmd (hunk 15 above)
Comment (profvjreddi) [reply frameworks.qmd:71]: This is one of the interesting things that I've been discovering about our notes here. And that is that when you start thinking about ML systems, be it tiny or otherwise, you still need to know the same set of fundamentals before you specialize in the domain of interest, such as tiny.

File: frameworks.qmd (hunk 9 above)
Comment (BaeHenryS) [reply frameworks.qmd:759]: Got it. thank you!

File: frameworks.qmd (hunk 12 above)
Comment (BaeHenryS) [reply frameworks.qmd:801]: Discussed this earlier. Thanks!

File: frameworks.qmd (hunk 21 above)
Comment (BaeHenryS) [reply frameworks.qmd:1139]: Got it. Thanks!
